### PR TITLE
ENH: Distances vs Positons v2

### PR DESCRIPTION
**Pull request by giovaniceotto**
## Pull request type

Please check the type of change your PR introduces:

- [x] Code base additions (bugfix, features)

## Pull request checklist

Please check if your PR fulfills the following requirements, depending on the type of PR:

- Code base additions (for bug fixes / features):

  - [ ] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [ ] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

See #158. While #195 solves #158, it has been merged to a branch which is currently stagnant. Therefore, it was refactored to be merged into the v1.0.0 release before that branch was reviewed.

## What is the new behavior?

RocketPy has been using distances between points since forever, but this has proven to be quite bothersome.
With this new implementation, the user can choose an arbitrary origin to his reference system and just report the position of all relevant points. Under the hood, RocketPy computes the necessary distances automatically.

This will certainly make it significantly easier to create rockets in RocketPy.

*The Motor class and its child classes have also gained their own coordinate system.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Other Information

Here is the current status of the tasks related to this PR. I have not modified the example notebooks and tests yet because I first want a review from @MateusStano to make sure that we agree on the implementation. Once the implementation is solid (which should happen after the first review), I will update unit tests and example notebooks.

### To open PR

- [x] Rocket class: eliminate use of distanceToCM
- [x] AeroSurfaces class : Update to handle new position attribute and reference system (lowest point).
- [x] Motor class: create new attributes (`centerOfMass` and `nozzlePosition`).
- [x] Flight class: replace `Rocket.distanceRocketNozzle` and `Rocket.distanceRocketPropellant`.

### After first PR review

- [x] Notebooks: update example notebooks to adhere to new system
- [x] Unit tests: update tests to new system
- [ ] Docs: add descriptive images with new reference system

### Cherry on top (perhaps another PR)

- [x] Create Rocket.addMotor
- [ ] Add support for multiple motors

## Points of Discussion

For this PR, I think the main points of discussion are:

- [x] Are the new parameter/attribute names good? How about their docstrings?
- [x] Does it make sense to give the user the freedom to also chose a reference system for Motors?
- [x] Should the axis really be positive upwards rather then downwards?
- [x] Are the reference points for AeroSurfaces sound?
- [x] Are the changes made to the CP calculation of AeroSurfaces correct?

### CODE REVIEW

**Comment by giovaniceotto** [comment]
# What has changed?

## Removed attributes

- **Rocket.distanceRocketNozzle** : float
    
    Distance between rocket's center of mass, without propellant, to the exit face of
    the nozzle, in meters. Always positive.

- **Rocket.distanceRocketPropellant** : float

    Distance between rocket's center of mass, without propellant, to the motor reference
    point, which for solid and hybrid motors is the center of mass of solid propellant,
    in meters. Always positive.

- **NoseCone.distanceToCM** : float

    Distance between nose cone tip and rocket center of mass. Has units of length and
    must be given in meters.

- **Tail.distanceToCM** : float

    Distance from the center of dry mass to the center of the tail.

- **Fins.position**, **TrapezoidalFins.position** and **EllipticalFins.position**: float

    Fin set position relative to rocket unloaded center of
    mass, considering positive direction from center of mass to
    nose cone.

## Removed arguments

- `Rocket.__init__`:
    * **distanceRocketNozzle** : int, float

        Distance from rocket's unloaded center of mass to nozzle outlet, in meters.
        Generally negative, meaning a negative position in the z axis which has an
        origin in the rocket's center of mass (without propellant) and points towards
        the nose cone.

    * **distanceRocketPropellant** : int, float
        Distance from rocket's unloaded center of mass to the motor reference point,
        which for solid and hybrid motor the is the center of mass of solid propellant,
        in meters. Generally negative, meaning a negative position in the z axis which
        has an origin in the rocket's center of mass (with out propellant) and points
        towards the nose cone.

- `Rocket.addTail`:
    * **distanceToCM** : int, float

        Tail position relative to rocket unloaded center of mass,
        considering positive direction from center of mass to nose
        cone. Consider the point belonging to the tail which is
        closest to the unloaded center of mass to calculate

- `Rocket.addNose`:
    * **distanceToCM** : int, float

        Nose cone position relative to rocket unloaded center of
        mass, considering positive direction from center of mass to
        nose cone. Consider the center point belonging to the nose
        cone base to calculate distance.

- `Rocket.addTrapezoidalFins` and `Rocket.addEllipticalFins`:
    *   **distanceToCM** : int, float

        Fin set position relative to rocket unloaded center of
        mass, considering positive direction from center of mass to
        nose cone. Consider the center point belonging to the top
        of the fins to calculate distance.

- `Rocket.setRailButtons`:
    *   **distanceToCM** : tuple, list, array
        
        Two values organized in a tuple, list or array which
        represent the distance of each of the two rail buttons
        to the center of mass of the rocket without propellant.
        If the rail button is positioned above the center of mass,
        its distance should be a positive value. If it is below,
        its distance should be a negative value. The order does
        not matter. All values should be in meters.

- `NoseCone.__init__`:
    * **distanceToCM** : float

        Distance between nose cone tip and rocket center of mass. Has units of
        length and must be given in meters.

- `Tail.__init__`:
    * **distanceToCM** : float

        Distance from the center of dry mass to the center of the tail.

- `Fins.__init__`, `TrapezoidalFins.__init__` and `EllipticalFins.__init__`:
    * **distanceToCM** : float

        Fin set position relative to rocket unloaded center of
        mass, considering positive direction from center of mass to
        nose cone. Consider the center point belonging to the top
        of the fins to calculate distance.

## New arguments

- `Rocket.__init__`:
    * **centerOfDryMassPosition** : int, float, optional

        Position of the rocket's center of dry mass (i.e. center of mass without
        propellant) relative to the user desired reference system. The only
        requirement is that the chosen reference system has a positive axis pointing
        from the rocket's tail to it's nose cone. The system's origin may be placed
        anywhere along such axis, such as in the nozzle or in the nose cone, and
        should be kept the same for all other positions specified later.
        Value expressed in meters. Default is 0, which means the center of dry mass
        is chosen as the origin, the legacy behavior of version 0.X.Y.
    
    * **motorPosition** : float

        Position of the motor's reference origin relative to the user defined rocket
        reference system. See Rocket.centerOfDryMassPosition for more information
        regarding the rocket reference system.

- `Rocket.addTail`:
    * **position** : int, float

        Tail position relative to user defined rocket reference system.
        By tail position, understand the point belonging to the tail which is 
        lowest in the rocket reference system (i.e. generally the point furthest
        from the nose cone).
        See `Rocket.centerOfDryMass` for more information regarding the rocket
        reference system.

- `Rocket.addNose`:
    * **position** : int, float

        Nose cone position relative to user defined rocket reference system.
        By nose cone position, understand the point belonging to the nose cone which
        is lowest in the rocket reference system (i.e. generally the center point
        of the nose cone base).
        See `Rocket.centerOfDryMass` for more information regarding the rocket
        reference system.`

- `Rocket.addTrapezoidalFins` and `Rocket.addEllipticalFins`:
    *   **position** : int, float
    
        Fin set position relative to user defined rocket reference system.
        By fin set position, understand the point belonging to the root chord which
        is lowest in the rocket reference system (i.e. generally the point furthest
        from the nose cone).
        See `Rocket.centerOfDryMass` for more information regarding the rocket

- `Rocket.setRailButtons`:
    *   **position** : tuple, list, array
        
        Two values organized in a tuple, list or array which
        represent the distance of each of the two rail buttons
        to the center of mass of the rocket without propellant.
        If the rail button is positioned above the center of mass,
        its distance should be a positive value. If it is below,
        its distance should be a negative value. The order does
        not matter. All values should be in meters.

- `Motor.__init__`, `SolidMotor.__init__` and `HybridMotor.__init__`:
    * **nozzlePosition** : int, float, optional

        Motor's nozzle outlet position in meters. More specifically, the coordinate
        of the nozzle outlet specified in a user defined reference frame which
        follows the motor's axis of symmetry, being positive upwards from the
        nozzle. The origin of such reference system maybe chosen by the user as
        desired.
        Default is 0, in which case the origin of the motor's reference system
        is placed at the motor's nozzle outlet.

- `SolidMotor.__init__`:
    * **grainsCenterOfMassPosition** : float

        Position of grains center of mass in meters. More specifically, the
        coordinate of the grains center of mass specified in a user defined
        reference frame which follows the motor's axis of symmetry, being
        positive upwards from the nozzle. The origin of such reference system
        maybe chosen by the user as desired, but must be consistent with the
        nozzlePosition argument.    

- `NoseCone.__init__`:
    * **position** : float

        Nose cone position relative to user defined rocket reference system.
        By nose cone position, understand the point belonging to the nose cone which
        is lowest in the rocket reference system (i.e. generally the center point
        of the nose cone base).
        See `Rocket.centerOfDryMass` for more information regarding the rocket
        reference system.

- `Tail.__init__`:
    * **position** : float

        Tail position relative to user defined rocket reference system.
        By tail position, understand the point belonging to the tail which is
        lowest in the rocket reference system (i.e. generally the point furthest
        from the nose cone).
        See `Rocket.centerOfDryMass` for more information regarding the rocket
        reference system.

- `Fins.__init__`, `TrapezoidalFins.__init__` and `EllipticalFins.__init__`:
    * **position** : int, float

        Fin set position relative to user defined rocket reference system.
        By fin set position, understand the point belonging to the root chord which
        is lowest in the rocket reference system (i.e. generally the point furthest
        from the nose cone).
        See `Rocket.centerOfDryMass` for more information regarding the rocket
        reference system.

## New attributes
- **Rocket.centerOfDryMassPosition** : int, float, optional

    Position of the rocket's center of dry mass (i.e. center of mass without
    propellant) relative to the user desired reference system. The only
    requirement is that the chosen reference system has a positive axis pointing
    from the rocket's tail to it's nose cone. The system's origin may be placed
    anywhere along such axis, such as in the nozzle or in the nose cone, and
    should be kept the same for all other positions specified later.
    Value expressed in meters. Default is 0, which means the center of dry mass
    is chosen as the origin, the legacy behavior of version 0.X.Y.

- **Rocket.motorPosition** : float

    Position of the motor's reference origin relative to the user defined rocket
    reference system. See Rocket.centerOfDryMassPosition for more information
    regarding the rocket reference system.

- **Rocket.centerOfPropellantPosition** : Function

    Position of the propellant's center of mass relative to the user defined
    rocket reference system. See Rocket.centerOfDryMassPosition for more
    information regarding the rocket reference system.
    Expressed in meters as a function of time.

- **SolidMotor.grainCenterOfMassPosition** : float

    Position of the center of mass of the grains in meters. More specifically,
    the coordinate of the center of mass specified in a user defined reference
    frame which follows the motor's axis of symmetry, being positive upwards
    from the nozzle. The origin of such reference system maybe chosen by the
    user as desired, but needs to be consistent with the Motor.nozzlePosition
    attribute.

- **SolidMotor.centerOfMass** : Function

    Position of the center of mass in meters as a function of time. Constant for solid
    motors, as the grains are assumed to be fixed.

- **Motor.nozzlePosition** : float

    Motor's nozzle outlet position in meters. More specifically, the coordinate
    of the nozzle outlet specified in a user defined reference frame which
    follows the motor's axis of symmetry, being positive upwards from the
    nozzle. The origin of such reference system maybe chosen by the user as
    desired.

- **NoseCone.position** : float

    Nose cone position relative to user defined rocket reference system.
    By nose cone position, understand the point belonging to the nose cone which
    is lowest in the rocket reference system (i.e. generally the center point
    of the nose cone base).
    See `Rocket.centerOfDryMass` for more information regarding the rocket
    reference system.

- **Tail.position** : float

    Tail position relative to user defined rocket reference system.
    By tail position, understand the point belonging to the tail which is
    lowest in the rocket reference system (i.e. generally the point furthest
    from the nose cone).
    See `Rocket.centerOfDryMass` for more information regarding the rocket
    reference system.

- **Fins.position**, **TrapezoidalFins.position** and **EllipticalFins.position**: float

    Fin set position relative to user defined rocket reference system.
    By fin set position, understand the point belonging to the root chord which
    is lowest in the rocket reference system (i.e. generally the point furthest
    from the nose cone).
    See `Rocket.centerOfDryMass` for more information regarding the rocket
    reference system.

## Modified Attributes

- **Rocket.centerOfMass** : Function

    Position of the rocket's center of mass, including propellant, relative
    to the user defined rocket reference system.
    See `Rocket.centerOfDryMassPosition` for more information regarding the
    coordinate system.            
    Expressed in meters as a function of time.

- **Rocket.cpPosition** : float

    Rocket's center of pressure position relative to the user defined rocket
    reference system. See `Rocket.centerOfDryMassPosition` for more information
    regarding the reference system.
    Expressed in meters.

## List of Motor Attributes used by the Rocket Class

```python
Motor.mass
Motor.thrust
Motor.burnOutTime
Motor.nozzlePosition
Motor.centerOfMass
```

## List of Motor Attributes used by the Flight Class

```python
Motor.nozzlePosition
Motor.nozzleRadius
Motor.thrust
Motor.burnOutTime
Motor.inertiaZ
Motor.inertiaI
Motor.inertiaZDot
Motor.inertiaIDot
Motor.mass
Motor.massDot
```

## List of Rocket Attributes used by the Flight Class

```python
Rocket.parachutes
Rocket.railButtons
Rocket.motor
Rocket.motorPosition
Rocket.centerOfDryMassPosition
Rocket.totalMass
Rocket.powerOnDrag
Rocket.area
Rocket.thrustEccentricityX
Rocket.thrustEccentricityY
Rocket.inertiaZ
Rocket.inertiaI
Rocket.mass
Rocket.powerOffDrag
Rocket.cpEccentricityY
Rocket.cpEccentricityX
Rocket.aerodynamicSurfaces
Rocket.radius
Rocket.reducedMass
Rocket.staticMargin
Rocket.thrustToWeight
Rocket.tipChord
Rocket.rootChord
Rocket.span
```

**Comment by giovaniceotto** [comment]
@MateusStano, as agreed, let's expedite the discussion and review of this PR so that it does not hold back the work that needs to be done for Liquid Motors and the v1.0.0 release.

**Comment by Gui-FernandesBR** [comment]
Tests not passing. See #281 first, @giovaniceotto

**Comment by Gui-FernandesBR** [comment]
Although it seems very cool, also contains several breaking changes.

I don't thing that develop should be the place for it. Instead, beta/v1.0.0 would be more safe

**Comment by giovaniceotto** [comment]
> 

I will review/approve that PR and rebase this one before this one is approved. Thanks for the reminder.

**Comment by giovaniceotto** [comment]
Consensus so far:

- Reference system positive axis can be chosen by the user (positive_towards_nose = True/False), same goes for motor
- Rocket.addMotor(test_motor, position)
- Aerodynamic surfaces reference position: closer to the nose cone

**Comment by giovaniceotto** [comment]
@MateusStano, I must admit, implement `addMotor` was much easier than I expected, but implementing the option to change coordinate system orientation took me forever.

It proved much harder than I had imagined mostly because there are so many reference systems at play (the one used in the Flight class, which cannot change since it must be right-handed, the one used in the Rocket class which the user can choose, the one being used in the motor class, which the user can also chose, and the ones used by the aerodynamic surfaces).

I believe I have made a good implementation overall. It is ready for your first review. I have uploaded a Getting Started notebook modified to work with the new code. It works! However, the static margin is strangely a little off. I have not been able to figure it out yet.

This review will definitely be really challenging, so I encourage @Gui-FernandesBR to help if possible since he reviewed the first PR related to this.

**Review by MateusStano — CHANGES_REQUESTED** [review]
I think we are only missing redoing all the tests here, after this I'd say we are all ready to merge.

Just to point out too, I think we need to change some things in `getting_started.ipynb` to better explain the reference points and such, but we can leave that for later

**Comment by MateusStano** [comment]
> However, the static margin is strangely a little off. I have not been able to figure it out yet.

I fixed it in the last commit, a line was missing in `evaluateStaticMargin`

**Comment by giovaniceotto** [comment]
> I fixed it in the last commit, a line was missing in `evaluateStaticMargin`

Wow, now everything makes sense... no wonder the static margin kept changing 😂.

@Gui-FernandesBR, I know it is a lot to ask, but if you could give me a hand modifying some unit tests, that would be really nice. Especially now that the liquid motor deadline is quite short (*and we do need this PR for the liquid motor PR)

File: rocketpy/Motor.py
Comment (giovaniceotto) [reply rocketpy/Motor.py:1779]: This will be improved with the LiquidMotor PR! For this PR, I left a simple implementation in order not to change many non-related lines of code.

**Comment by giovaniceotto** [comment]
Alright @MateusStano and @Gui-FernandesBR. We need to merge this by today so that we can move forward with #233.

**Comment by Gui-FernandesBR** [comment]
> Alright @MateusStano and @Gui-FernandesBR. We need to merge this by today so that we can move forward with #233.

@giovaniceotto , as this is not a simple "search and replace" task, I was afraid of commiting a lot of stuffs directly into your branch, therefore I'll be working on `tst/adjusting_test_distances_vs_positions` branch and I let you know once I'm done.

This you be ready by today, as requested. I'll try my best but if (actually when) I need help, I'll go directly to @MateusStano , don't worry.

**Comment by Gui-FernandesBR** [comment]
Description of `[26e4fd4](https://github.com/RocketPy-Team/RocketPy/pull/282/commits/26e4fd4eedb066530b1e35a2991fd0ec5b2237de)` : 

I had to-do that to upstream current head branch with tools.py, otherwise it would be impossible to load cached_property module  in python 3.7

**Comment by Gui-FernandesBR** [comment]
> Description of `[26e4fd4](https://github.com/RocketPy-Team/RocketPy/pull/282/commits/26e4fd4eedb066530b1e35a2991fd0ec5b2237de)` :
> 
> I had to-do that to upstream current head branch with tools.py, otherwise it would be impossible to load cached_property module in python 3.7

OK, everything is done and waiting for your review/approval, @giovaniceotto , please check #305 PR

**Comment by giovaniceotto** [comment]
All right, here are the few things missing for this PR to be merged:

- [ ] Add tests to make sure `noseToTail` and `combustionChamberToNozzle` coordinate system orientations work as expected.
- [ ] Update notebooks to new coordinate system.
- [ ] Improve docs by adding images regarding the coordinate system and their orientation.

The core implementation itself is in its final form, so if you are working on something that depends on this, go ahead and merge this branch as is into your working branch. It should not cause problems in the future.

**Comment by MateusStano** [comment]
Okay, everything is done now! Here is what I did:

- Added `remove` and `pop` methods to `class AeroSurfaces` because they were needed in some of the analysis.

- Created a test in test_rocket.py to check different coordinate system orientations

- Added a lot of explanations on the getting started notebooks plus some images. If you open the notebook right now the images wont show up because they are linked to their github link whenever they go up to master. I don't know if there is a better way to link images in notebooks though. Here are the images explaining the coordinate systems:

motorOrientation

calistoMotorReference


rocketCsys


calistoCsys

**Review by MateusStano — APPROVED** [review]

**Comment by giovaniceotto** [comment]
Amazing work @MateusStano, just amazing!

I'll take a deeper look into this by tomorrow and see if I can find another way of referencing the images.

File: rocketpy/AeroSurfaces.py
Comment (giovaniceotto) [inline rocketpy/AeroSurfaces.py:30]: I am quite confused about this method. The list `pop` method takes a single input, the index of the item to be removed, but here, a tuple is given. By default, if no index is given, -1 is used (i.e. the last item of the list is removed and returned).

Here is a possible implementation:
```suggestion
    def pop(self, index=-1):
        return self._aerosurfaces.pop(index)
```

This way, the method can be used to remove the last aerodynamic surface added. Since it is returned, it can be modified and later on be added or appended again.

File: tests/test_rocket.py
Comment (giovaniceotto) [inline tests/test_rocket.py:165]: I quite like this assertion. Simple, but effective in catching implementation errors.

File: tests/test_rocket.py
Comment (giovaniceotto) [inline tests/test_rocket.py:23]: Sorry, why was this change needed?

**Comment by giovaniceotto** [comment]
@MateusStano, how did you manage to update the MATLAB Getting Started html file? Did you edit the livescript and generate the html again?

**Comment by giovaniceotto** [comment]
@MateusStano, do you have the "source" for the png images you created regarding the coordinate system?

File: tests/test_rocket.py
Comment (MateusStano) [reply tests/test_rocket.py:23]: Oh, it wasnt needed. Every other motor definition on the test_rocket.py is also like this. Will fix it soon

**Comment by MateusStano** [comment]
> @MateusStano, do you have the "source" for the png images you created regarding the coordinate system?

I added them to our gdrive. I also managed to link the notebook to the images on the drive, so now they should be showing up in the notebooks

**Comment by MateusStano** [comment]
> @MateusStano, how did you manage to update the MATLAB Getting Started html file? Did you edit the livescript and generate the html again?

I opened the HTML on my browser and edited it in there. Smart use of ctrl+f definitely helped


> If so, what are your toughts on using RocketPy in MATLAB? And how can it be improved?

I actually don't know much about it, never had even interfaced with it before editing the .html and .m files. I also never used anything like it in MATLAB so I really don't have much feedback to give

**Comment by Gui-FernandesBR** [comment]
I just loved the new drawings for further explanations, this will help us a lot! 

I am mentioning here an issue that can be leveraged by these new implementation: #159 

Perhaps in the future we can drop the gdrive links dependency on the notebooks, but for now it seems a clever solution, nice!

**Review by Gui-FernandesBR — APPROVED** [review]
I only would suggest running black on the notebooks as well. But not mandatory.

Very very nice PR.

**Comment by giovaniceotto** [comment]
@MateusStano, where the images created using a vector base platform, such as PowerPoint/Inkscape or other?

**Comment by MateusStano** [comment]
> @MateusStano, were the images created using a vector base platform, such as PowerPoint/Inkscape or other?

Yeah, It is on our team Figma. Just sent you an invite

**Comment by Gui-FernandesBR** [comment]
Hey, the #306 is coming towards to this branch, and it is already approved. 
I'd love to merge it here, but first I'd like to make sure this PR is approved and finished.
Therefore, I'm waiting for @giovaniceotto 's review.

Please do not merge this one into beta/v1.0.0 without the changes of #306

**Comment by Gui-FernandesBR** [comment]
Tests stopped working since last merge commit. 

@MateusStano, do you think you could solve it?

**Comment by MateusStano** [comment]
It is the same problem we had in #306 

> Great work! I changed some very minor things. The biggest thing was a small bug inside the Function class that returned nan when receiving np.inf in the __call__ method in int or float defined Functions.

I think by merging that one into this should fix things

**Comment by giovaniceotto** [comment]
The MATLAB script is not running properly in my system. Close this PR and I will solve it in another one. How does that sound?

**Comment by Gui-FernandesBR** [comment]
> The MATLAB script is not running properly in my system. Close this PR and I will solve it in another one. How does that sound?

Sounds good to me as long as this correction comes earlier than the v1.0.0 release.
So, sure, sounds totally good.

Also, let's focus on the main/core features now, the MATLAB integration can be fixed later

**Pull request merged**
